File: backend/internal/utils/money.go
package utils

import (
	"fmt"
	"strconv"
	"strings"
)

// Money in this application is always an int64 count of minor units -- paise for INR.
// No float64 ever touches an amount (DECISIONS.md D7). These helpers are the only place
// minor units are converted for display, and they are exact.

// BasisPointsDivisor converts a basis-point rate to a fraction: 500 bps = 5.00%.
const BasisPointsDivisor = 10000

// ApplyBasisPoints returns amount * bps / 10000, rounded half-up.
//
// Rounding is explicit because integer division truncates, and truncating tax
// systematically undercharges by up to one paisa per order -- which is both wrong and,
// aggregated over a year of service, a real number on a GST return.
func ApplyBasisPoints(amountMinor int64, bps int) int64 {
	if bps <= 0 || amountMinor == 0 {
		return 0
	}
	product := amountMinor * int64(bps)
	// Half-up: add half the divisor before dividing.
	return (product + BasisPointsDivisor/2) / BasisPointsDivisor
}

// FormatMinor renders minor units as a plain decimal string, "24950" -> "249.50".
func FormatMinor(amountMinor int64) string {
	neg := amountMinor < 0
	if neg {
		amountMinor = -amountMinor
	}
	major := amountMinor / 100
	minor := amountMinor % 100
	s := fmt.Sprintf("%d.%02d", major, minor)
	if neg {
		return "-" + s
	}
	return s
}

// FormatINR renders minor units for display with the rupee symbol and Indian digit
// grouping: 1234567890 paise -> "Rs 1,23,45,678.90".
//
// Indian grouping is not the western thousands separator -- it groups the last three
// digits, then pairs. A bill that reads "12,345,678" to an Indian diner looks like a
// typo, and this is a diner-facing product.
func FormatINR(amountMinor int64) string {
	neg := amountMinor < 0
	if neg {
		amountMinor = -amountMinor
	}
	major := strconv.FormatInt(amountMinor/100, 10)
	minor := fmt.Sprintf("%02d", amountMinor%100)

	grouped := groupIndian(major)
	if neg {
		return "-₹" + grouped + "." + minor
	}
	return "₹" + grouped + "." + minor
}

// groupIndian inserts separators using the Indian lakh/crore convention.
func groupIndian(n string) string {
	if len(n) <= 3 {
		return n
	}
	head, tail := n[:len(n)-3], n[len(n)-3:]

	var parts []string
	for len(head) > 2 {
		parts = append([]string{head[len(head)-2:]}, parts...)
		head = head[:len(head)-2]
	}
	if head != "" {
		parts = append([]string{head}, parts...)
	}
	return strings.Join(parts, ",") + "," + tail
}

// ParseMajorToMinor converts a decimal string such as "249.50" into minor units.
//
// Used only at trusted admin-input boundaries (menu price entry). It rejects more than
// two decimal places rather than rounding, so a mistyped "249.555" is a visible error
// instead of a silently altered price.
func ParseMajorToMinor(s string) (int64, error) {
	s = strings.TrimSpace(strings.ReplaceAll(s, ",", ""))
	if s == "" {
		return 0, fmt.Errorf("empty amount")
	}

	neg := strings.HasPrefix(s, "-")
	s = strings.TrimPrefix(strings.TrimPrefix(s, "-"), "+")

	whole, frac, hasFrac := strings.Cut(s, ".")
	if whole == "" {
		whole = "0"
	}
	major, err := strconv.ParseInt(whole, 10, 64)
	if err != nil {
		return 0, fmt.Errorf("invalid amount %q", s)
	}

	var minor int64
	if hasFrac {
		if len(frac) > 2 {
			return 0, fmt.Errorf("amount %q has more than two decimal places", s)
		}
		padded := (frac + "00")[:2]
		if minor, err = strconv.ParseInt(padded, 10, 64); err != nil {
			return 0, fmt.Errorf("invalid amount %q", s)
		}
	}

	total := major*100 + minor
	if neg {
		total = -total
	}
	return total, nil
}

File: packages/shared/src/money.ts
/**
 * Money helpers.
 *
 * The server sends both the integer paise and a pre-formatted display string, so the
 * client should normally render `money.display` and do no arithmetic at all
 * (docs/DECISIONS.md D7). These exist for the one case the server cannot cover: a cart
 * total recomputed locally as the diner taps + and -, before anything is sent.
 *
 * Every function here works in integer paise. No float ever touches an amount.
 */

/** Basis points divisor: 500 bps = 5.00%. */
export const BPS_DIVISOR = 10_000

/**
 * Applies a basis-point rate, rounding half-up.
 *
 * Rounding is explicit because truncation systematically undercharges by up to a paisa per
 * order, which is both wrong and a real number on a year's GST return. Matches
 * `utils.ApplyBasisPoints` in the Go backend exactly, so a locally computed cart total and
 * the server's authoritative total agree to the paisa.
 */
export function applyBasisPoints(amountMinor: number, bps: number): number {
  if (bps <= 0 || amountMinor === 0) return 0
  const product = amountMinor * bps
  return Math.floor((product + BPS_DIVISOR / 2) / BPS_DIVISOR)
}

/**
 * Formats paise for display with Indian digit grouping: 12345678 -> "₹1,23,456.78".
 *
 * Indian grouping is not the western thousands separator -- it groups the last three digits
 * then pairs. "12,345,678" reads as a typo to an Indian diner, and this is a diner-facing
 * product. Mirrors `utils.FormatINR` in the Go backend.
 */
export function formatINR(amountMinor: number): string {
  const negative = amountMinor < 0
  const abs = Math.abs(Math.trunc(amountMinor))

  const major = Math.floor(abs / 100).toString()
  const minor = (abs % 100).toString().padStart(2, '0')

  return `${negative ? '-' : ''}₹${groupIndian(major)}.${minor}`
}

/** Inserts separators using the lakh/crore convention. */
function groupIndian(digits: string): string {
  if (digits.length <= 3) return digits

  const tail = digits.slice(-3)
  let head = digits.slice(0, -3)

  const parts: string[] = []
  while (head.length > 2) {
    parts.unshift(head.slice(-2))
    head = head.slice(0, -2)
  }
  if (head) parts.unshift(head)

  return `${parts.join(',')},${tail}`
}

/** Formats paise without a currency symbol: 24950 -> "249.50". */
export function formatMinor(amountMinor: number): string {
  const negative = amountMinor < 0
  const abs = Math.abs(Math.trunc(amountMinor))
  return `${negative ? '-' : ''}${Math.floor(abs / 100)}.${(abs % 100).toString().padStart(2, '0')}`
}

/** A locally computed price breakdown, for the cart before checkout. */
export interface ComputedTotals {
  subtotalMinor: number
  taxMinor: number
  serviceChargeMinor: number
  totalMinor: number
}

/**
 * Recomputes a cart total client-side, for instant feedback as quantities change.
 *
 * This is a display convenience, never the authority: the server re-prices the whole order
 * from the live menu at placement, which is why `PlaceOrderRequest` carries no amount at
 * all. Kept arithmetically identical to the Go implementation so the number the diner sees
 * in the cart is the number on their bill.
 */
export function computeTotals(
  lines: ReadonlyArray<{ unitPriceMinor: number; quantity: number }>,
  taxBps: number,
  serviceChargeBps: number,
): ComputedTotals {
  const subtotalMinor = lines.reduce((sum, l) => sum + l.unitPriceMinor * l.quantity, 0)
  const taxMinor = applyBasisPoints(subtotalMinor, taxBps)
  const serviceChargeMinor = applyBasisPoints(subtotalMinor, serviceChargeBps)

  return {
    subtotalMinor,
    taxMinor,
    serviceChargeMinor,
    totalMinor: subtotalMinor + taxMinor + serviceChargeMinor,
  }
}

/** Renders a basis-point rate as a percentage label: 500 -> "5%", 250 -> "2.5%". */
export function formatBps(bps: number): string {
  const percent = bps / 100
  return `${Number.isInteger(percent) ? percent : percent.toFixed(2).replace(/0$/, '')}%`
}
